src/frontend/xgboost_json.cc

```diff
@@ -25,45 +43,363 @@ namespace frontend {
+                                             split_conditions) ||
+      push_key_handler<ArrayHandler<bool>>("default_left", default_left));
+}
+
+bool RegTreeHandler::StartObject() {
+  return push_key_handler<TreeParamHandler, int>("tree_param", num_nodes);
+}
 
-  while (true) {
-    str.resize(total_read + size);
-    size_t read = fi->Read(&str[total_read], size);
-    total_read += read;
-    if (read < size) {
-      break;
+bool RegTreeHandler::Uint(unsigned u) { return check_cur_key("id"); }
+
+bool RegTreeHandler::EndObject(std::size_t memberCount) {
+  m_output.Init();
+  if (num_nodes != loss_changes.size() || num_nodes != sum_hessian.size() ||
+      num_nodes != base_weights.size() ||
+      num_nodes != leaf_child_counts.size() ||
+      num_nodes != left_children.size() ||
+      num_nodes != right_children.size() || num_nodes != parents.size() ||
+      num_nodes != split_indices.size() ||
+      num_nodes != split_conditions.size() ||
+      num_nodes != default_left.size()) {
+    return false;
+  }
+
+  std::queue<std::pair<int, int>> Q;  // (old ID, new ID) pair
+  Q.push({0, 0});
+  while (!Q.empty()) {
+    int old_id, new_id;
+    std::tie(old_id, new_id) = Q.front();
+    Q.pop();
+
+    if (left_children[old_id] == -1) {
+      m_output.SetLeaf(new_id, base_weights[old_id]);
+    } else {
+      m_output.AddChilds(new_id);
+      m_output.SetNumericalSplit(
+          new_id, split_indices[old_id], split_conditions[old_id],
+          default_left[old_id], treelite::Operator::kLT);
+      m_output.SetGain(new_id, loss_changes[old_id]);
+      Q.push({left_children[old_id], m_output.LeftChild(new_id)});
+      Q.push({right_children[old_id], m_output.RightChild(new_id)});
     }
-    size *= 2;
+    m_output.SetSumHess(new_id, sum_hessian[old_id]);
   }
-  str.resize(total_read);
+  return pop_handler();
+}
+
+/******************************************************************************
+ * GBTreeHandler
+ * ***************************************************************************/
+bool GBTreeModelHandler::StartArray() {
+  return (push_key_handler<ArrayHandler<treelite::Tree, RegTreeHandler>,
+                           std::vector<treelite::Tree>>("trees",
+                                                        m_output.trees) ||
+          push_key_handler<IgnoreHandler>("tree_info"));
+}
 
-  return str;
+bool GBTreeModelHandler::StartObject() {
+  return push_key_handler<IgnoreHandler>("gbtree_model_param");
 }
 
+/******************************************************************************
+ * GradientBoosterHandler
+ * ***************************************************************************/
+bool GradientBoosterHandler::Uint(unsigned u) {
+  return check_cur_key("num_boosting_round");
+}
+bool GradientBoosterHandler::String(const char *str,
+                                    std::size_t length,
+                                    bool copy) {
```

Comment:
```suggestion
                                    bool copy) {
  if (!check_cur_key("name")) {
    return false;
  }
```
We should first assert that the key is equal to "name".